Walkthrough 3 - low code development (#92)
integr8ly/tutorial-web-app-walkthroughs: walkthroughs/3a-low-code-api-development/walkthrough.adoc

// tag::master-1[]

:walkthrough: No code API development and SaaS integration
:fuse-version: 7.2
:3scale-name: Red Hat 3scale API Management Platform
:3scale-version: 7.2
:fuse-url: https://eval.apps.city.openshiftworkshop.com/
:3scale-url: https://eval.apps.city.openshiftworkshop.com/
:route: https://wt2-{user-sanitized-username}-3scale.{openshift-app-host}

[id='no-code-container-native-api-development']

= {walkthrough}

> This walkthrough demonstrates how to implement a container-native OpenAPI RESTful API without coding.

Reviewer: Formatting looks a bit iffy.  See https://www.dropbox.com/s/s6jfzqduojpk1mq/Screenshot%202019-08-23%20at%2011.46.27.png?dl=0

Reviewer 2: create a jira

Reviewer: Done.  https://issues.jboss.org/browse/INTLY-3067

Container-native applications::
Applications that are designed to run in a containerised environment.

// This is taken right from https://github.com/OAI/OpenAPI-Specification
OpenAPI Specification::

> A standard, programming language-agnostic interface description for REST APIs, which allows both humans and computers to discover and understand the capabilities of a service without requiring access to source code, additional documentation, or inspection of network traffic. When properly defined via OpenAPI, a consumer can understand and interact with the remote service with a minimal amount of implementation logic.

Reviewer: Is this sentence too long?

Reviewer 2: make suggestion

Reviewer: A standard, programming language-agnostic interface description for REST APIs. It allows both humans and computers to discover and understand the capabilities of a service without requiring access to source code, additional documentation, or inspection of network traffic.

It also allows you to identify important design decisions prior to coding and deployment.

[type=walkthroughResource,serviceName=fuse]
.Fuse Online
****
* link:{fuse-url}[Console, window="_blank"]
****

[type=walkthroughResource,serviceName=openshift]
.Red Hat OpenShift
****
* link:{openshift-host}/console[Console, window="_blank"]
****

[type=walkthroughResource,serviceName=3scale]
.3Scale
****
* link:{api-management-url}[Console, window="_blank"]
* link:https://developers.redhat.com/products/3scale/overview/[3Scale Overview, window="_blank"]
* link:https://www.3scale.net[3Scale Website, window="_blank"]
****

[time=10]
== Creating a Slack Connection in Fuse Online

In Fuse Online a *Connection* holds the information required to connect to a service.
For example, a Slack *Connection* in Fuse Online requires an *API Token* and *Slack URL* to function. 

=== Creating a Slack Workspace

.Prerequisite
You already have an account with link:https://slack.com/[Slack].

.Procedure
. Navigate to the link:https://slack.com/create[Create a new workspace, window="_blank"] page.

. Enter your email address and click *Next*.

> 

Reviewer: If you have 2FA enabled on your slack account, you will have to enter the 2FA code after clicking *Next*.  It won't go straight to "What's the name of your company or team?"

> . When prompted *What's the name of your company or team'* enter::

Reviewer: ```suggestion
. When prompted *What's the name of your company or team?* enter::
```

+
----
low-code-workspace
----

. When prompted *What’s a project your team is working on?* enter:
+
----
low-code-demo
----

. Click *Skip for now* when prompted to add team members.

> 

Reviewer: ```suggestion
. Click *See Your Channel in Slack*
```

[type=verification]
Can you access your workspace, and navigate to the *low-code-demo* channel listed on the left of the workspace?

[type=verificationFail]
Verify that you followed each step in the procedure above.  If you are still having issues, check your email, there may be information sent to you about accessing your workspace.

=== Creating the Slack Connection in Fuse Online

> . Log in to the link:{fuse-url}[Red Hat Fuse Online, window="_blank"] console.

Reviewer: {fuse-url} points to https://eval.apps.city.openshiftworkshop.com/ , which is a dead link.  Unable to proceed further.

+
NOTE: This is your personal Fuse Online instance and not the shared instance available from link:/[home page].

. Choose *Connections* from the left navigation menu.

. Click the *Create Connection* button in the top right corner.

. Choose *Slack* from the list of displayed *Connectors*. 
You are prompted for configuration details.

. Obtain a URL for the *Slack Webhook URL* field:
.. Navigate to the Slack Apps portal at link:https://api.slack.com/apps[api.slack.com/apps].

> .. Login using the email address associated with your workspace.

Reviewer: You should be already logged into slack at this stage.

.. Click *Create new App*.
.. Enter the following in the *App Name* field:
+
----
Fuse Online Demo
----
.. Choose *low-code-workspace* from the *Development Slack Workspace* menu.
.. Click "Create App".
.. On the following screen under the *Add features and functionality* click *Incoming Webhooks*.
.. Change the *Activate Incoming Webhooks* switch to the *On* position.
.. Scroll down and click *Add New Webhook to Workspace*.
.. In the *Post to* field select *low-code-demo* and click *Allow*.
+
You are redirected back to the *Incoming Webhooks* page. 
.. Copy the *Webhook URL* displayed at the bottom of the page using the *Copy* button. 

. Paste the copied *Webhook URL* value into the *Slack Webhook URL* field in Fuse Online.

. Obtain a value for the *Token for Accessing Slack API* field:
.. Navigate to the Slack API Token portal at link:https://api.slack.com/custom-integrations/legacy-tokens[api.slack.com/custom-integrations/legacy-tokens, window="_blank"].

> .. Login using the email address associated with your workspace.

Reviewer: Logged in already.

.. Scroll down to the *Legacy token generator* section.

> .. Click *Create token* button next to a workspace you own, or have admin rights for.

Reviewer: ```suggestion
.. Click *Create token* button next to *low-code-workspace*.
```

. Enter the following in the *Sending User Name for Messages* field:
+
----
Fuse Online
----

. Click *Validate* in Fuse Online and verify that a *Slack has been successfully validated.* message appears. Click *Next*.

. Enter the following in the *Connection Name* field on the *Add Connection Details* screen:
+
----
Greeting Slack Target
----

. Click *Save*.

[type=verification]
Is a *Greeting Slack Target* entry listed in the Fuse Online *Connections* screen?

[type=verificationFail]
Verify that you followed each step in the procedure above.  If you are still having issues, contact your administrator.

[type=verificationFail]
Verify that you followed each step in the procedure above.  If you are still having issues, contact your administrator.

[time=10]
== Creating an API Integration in Fuse Online

=== Creating an Integration with API details

In this procedure, you create an integration with an API and include a data type.
A *Data Type* represents a data structure that can be passed to your API and can then be used as parameters in your API definition and within the Fuse Online *Integration Flow Editor* to transform and map data as part of an *Integration*.

> 

Reviewer: ```suggestion

. Choose *Integrations* from the left navigation menu.

```

. Click the *Create Integration* button to start the *New Integration* wizard.

. Choose *API Provider* on the subsequent *Choose a Start Connection* screen.

. When prompted choose *Create* and click *Next* to navigate to the *API Designer*.

. Rename your API from "Untitled API" to:
+
----
Greeting API
----

. Edit the *Description*:
+
----
My greeting API
----

. Click *Add a data type* under the *Data Types* heading on the left of the *API Designer* screen.

. In the *Enter Basic Information* section enter the following in the *Name* field:
+
----
Name
----

. Enter the following JSON in the *Enter JSON Example* field:
+
[subs="attributes+"]
----
{

>     "name": "shadowman" 

Reviewer: ```suggestion
"name": "shadowman" 
```

Reviewer 3: I would leave this as-is; indentation is expected in the json object.

}
----

. Scroll down and click *Save*.

[type=verification]
Is `</> Name` listed under the *Data Types* on the *API Designer* screen?

[type=verificationFail]
Verify that you followed each step in the procedure above.  If you are still having issues, contact your administrator.

=== Creating a POST Resource Path

A *Path* represents an API endpoint/operation and the associated parameters required to invoke it.

. Click *Add a path* under the *Paths* heading on the left of the *API Designer* screen.

. In the window that appears, enter the following in the *Path* field:
+
----
/greeting
----

. Click *Add* to confirm your entry.

> . Click your new */greeting* endpoint under the *Paths* heading.

Reviewer: No need for this instruction, already clicked.

. Click *Add Operation* under the *POST* icon in the *Operations* section on the right.

. Select *POST* from the Operations heading.

. Enter the following in the *Summary* field:
+
----
Greet with name
----

. Enter the following in the *Operation ID* field:
+
----
greetname
----

> . In the *Request Body* section, choose the *Name* type that you created earlier.

Reviewer: Needs more details.  It is not possible to choose the *Name* data type as "No request body has been defined.".  Type should be "string as string".

Reviewer 2: out of date

. In the *Responses* section click *Add response*.
.. In the window that appears choose *200 OK* as the dropdown option.
.. Click *Add*.
.. Click *No description* beside the *200 OK* response and enter the following in the *Description* field:
+
----
Greeting response
----

. Click *Save* in the top right corner of the UI to save your work and be directed back to the *New Integration* wizard.

> . Click *Publish* on the *New Integration* wizard screen.

Reviewer: There is no "publish" button.  Clicking "Next" brings you to a screen with a "Publish" button.

. When prompted to *Give this integration a name*:
.. Enter the following in the *Integration Name* field:
+
----
Greeting Integration
----
.. Enter the following in the *Description* field:
+
----
my greeting api
----

> . Click *Save* to save your API design.

Reviewer: ```suggestion
. Click *Save and Publish* to save your API design.
```

Reviewer 2: done

. Click the blue plus icon to add a step.
The left hand side of the *Choose a connection* page lists the steps in your *Integration*.

. Select your *Greeting Slack Target*.

. When prompted to *Choose an action* select *Channel*. You can use this to send a message to a specific channel in your Slack workspace.

. Use the *Channel* dropdown to select the *low-code-demo* channel and click *Next*.
The left hand side of the *Integration Flow Editor* should now list your Slack connection with a *Data Type Mismatch* warning. 

. Click the *Warning Icon* and choose *Add a data mapping step*.

. From the *Configure Mapper* screen click the *body* element in the *Source* panel to expand it.

. Click the *name* field under the *body*, then click the *message* element in the *Target* panel. This maps the value of the incoming HTTP request *body* to the outgoing Slack *message* property.

. In the right hand *Mapping Details* panel, click the *Arrow Icon* under the *Targets* section to add a transformation.

. Using the dropdown change the transformation type from *Append* to *Prepend*.

. Enter the following in the *string* field under the dropdown:
+
----
Hello from 
----

. Click *Done*.

> . Click *Publish* to trigger an link:https://docs.openshift.com/container-platform/3.11/creating_images/s2i.html[Source to Image (S2I), window="_blank"] build.

Reviewer: ```suggestion
. Click *Publish*, and then *Save and Publish", to trigger an link:https://docs.openshift.com/container-platform/3.11/creating_images/s2i.html[Source to Image (S2I), window="_blank"] build.
```

Reviewer 2: done

+
You are redirected back to the *Incoming Webhooks* page. 
.. Copy the *Webhook URL* displayed at the bottom of the page using the *Copy* button. 

. Paste the copied *Webhook URL* value into the *Slack Webhook URL* field in Fuse Online.

. Obtain a value for the *Token for Accessing Slack API* field:
.. Navigate to the Slack API Token portal at link:https://api.slack.com/custom-integrations/legacy-tokens[api.slack.com/custom-integrations/legacy-tokens, window="_blank"].
.. Login using the email address associated with your workspace.
.. Scroll down to the *Legacy token generator* section.
.. Click *Create token* button next to a workspace you own, or have admin rights for.
.. Copy the generated token.

. Paste the generated token into the *Token for Accessing Slack API* field in Fuse Online.

. Enter the following in the *Sending User Name for Messages* field:
+
----
Fuse Online
----

. Click *Validate* in Fuse Online and verify that a *Slack has been successfully validated.* message appears. Click *Next*.

. Enter the following in the *Connection Name* field on the *Add Connection Details* screen:
+
----
Greeting Slack Target
----

> . Click *Create*.

Reviewer 3: ```suggestion
. Click *Save*.
```

[type=verification]
Is a *Greeting Slack Target* entry listed in the Fuse Online *Connections* screen?

[type=verificationFail]
Verify that you followed each step in the procedure above.  If you are still having issues, contact your administrator.

[type=verificationFail]
Verify that you followed each step in the procedure above.  If you are still having issues, contact your administrator.

[time=10]
== Creating an API Integration in Fuse Online

=== Creating an Integration with API details

In this procedure, you create an integration with an API and include a data type.
A *Data Type* represents a data structure that can be passed to your API and can then be used as parameters in your API definition and within the Fuse Online *Integration Flow Editor* to transform and map data as part of an *Integration*.

. Click the *Create Integration* button to start the *New Integration* wizard.

. Choose *API Provider* on the subsequent *Choose a Start Connection* screen.

> . When prompted choose *Create from scratch* and click *Next* to navigate to the *API Designer*.

Reviewer 3: ```suggestion
. When prompted choose *Create* and click *Next* to navigate to the *API Designer*.
```

=== Creating a POST Resource Path

A *Path* represents an API endpoint/operation and the associated parameters required to invoke it.

. Click *Add a path* under the *Paths* heading on the left of the *API Designer* screen.

. In the window that appears, enter the following in the *Path* field:
+
----
/greeting
----

. Click *Add* to confirm your entry.

. Click your new */greeting* endpoint under the *Paths* heading.

. Click *Add Operation* under the *POST* icon in the *Operations* section on the right.

. Select *POST* from the Operations heading.

. Enter the following in the *Summary* field:
+
----
Greet with name
----

. Enter the following in the *Operation ID* field:
+
----
greetname
----

. In the *Request Body* section, choose the *Name* type that you created earlier.

. In the *Responses* section click *Add response*.
.. In the window that appears choose *200 OK* as the dropdown option.
.. Click *Add*.
.. Click *No description* beside the *200 OK* response and enter the following in the *Description* field:
+
----
Greeting response
----

. Click *Save* in the top right corner of the UI to save your work and be directed back to the *New Integration* wizard.

> . Click *Next* on the *New Integration* wizard screen.

Reviewer 3: ```suggestion
. Click *Publish* on the *New Integration* wizard screen.
```

> 

Reviewer 3: ```suggestion
. Click *Create flow* for the /greeting POST operation you just created
```

. When prompted to *Give this integration a name*:
.. Enter the following in the *Integration Name* field:
+
----
Greeting Integration
----
.. Enter the following in the *Description* field:
+
----
my greeting api
----

> . Click *Save and Continue* to save your API design.

Reviewer 3: ```suggestion
. Click *Save* to save your API design.
```

[type=verification]
Is a *Greeting Integration* entry listed in the Fuse Online Home screen?

[type=verificationFail]
Verify that you followed each step in the procedure above.  If you are still having issues, contact your administrator.

[time=10]
== Implementing and publishing the API

. Log in to the link:{fuse-url}[Red Hat Fuse Online, window="_blank"] console.

. Select *Integrations* from the left hand menu. 

. Choose *Edit* from the menu for the *Greeting Integration* entry. The *Operations* page appears.

> . Click the *Edit flow* button. The *Add to Integration* page appears displaying the *Provided API* and the *Provided API Return Path*.

Reviewer 3: ```suggestion
. Click the *Create flow* button. The *Add to Integration* page appears displaying the *Provided API* and the *Provided API Return Path*.
```

. Click the blue plus icon to add a step.
The left hand side of the *Choose a connection* page lists the steps in your *Integration*.

. Select your *Greeting Slack Target*.

. When prompted to *Choose an action* select *Channel*. You can use this to send a message to a specific channel in your Slack workspace.

> . Use the *Channel* dropdown to select the *low-code-demo* channel and click *Done*.

Reviewer 3: ```suggestion
. Use the *Channel* dropdown to select the *low-code-demo* channel and click *Next*.
```

=== API Management Login

. Open the link:{api-management-url}[{3scale-name} Login screen, window="_blank"].

. Select the *Red Hat Single Sign On* option. This triggers an OAuth Flow and redirects you back to the {3scale-name} Dashboard.

. Dismiss the *How does 3Scale work?* option which is displayed the first time you log in to {3scale-name}. The main Dashboard is displayed.

[type=verification]
Can you see the {3scale-name} Dashboard and navigate the main menu?

[type=verificationFail]
Verify that you followed each step in the procedure above.  If you are still having issues, contact your administrator.

=== Adding the App Endpoint to Red Hat 3scale

. From the *Dashboard*, select the *New API* item.
. Select the *Import from OpenShift* option.
If this option is not enabled, click *authenticate with OpenShift* to enable the option.

. Choose the 'fuse' option from the *Namespace* list.
. Choose 'i-greeting-api' from the *Name* list.
. Click *Create Service*.
+
This process can take a few minutes.

// The 3scale dashboard is displayed, with notification: `The service will be imported shortly. You will receive a notification when it is done.`

. Edit the API:

> .. From the *Overview* screen, select the *Configure APIcast* button.

Reviewer 3: ```suggestion
.. From the *Overview* screen, select the *edit APIcast configuration* link.
```

----
.. Leave the other fields with their default values.
.. Select *Create Application Plan*. You will be redirected to the *Application Plans* screen.
.. Select the *Publish* button, beside your plan list item, to publish the Plan.

. Create a new *Application* for the *Developer* Group, assigned to the Plan:
.. Select *Audience* from the top navigation bar dropdown.
.. Select the *Developer* Account to open the *Account Summary* page.
.. Select the *(num) Application* item from the breadcrumb to view Applications.
.. Select the *Create Application* button in the top right.
.. Select the *low-code* Plan in the *Application plan* dropdown.
.. Enter the following for *Name* and *Description*:
+
[subs="attributes+"]
----
low-code-app
----
.. Select *Create Application*.

. Set a custom *User Key* for the application:
.. On the *low-code* application screen you were redirected to, scroll to the *API Credentials* section.
.. Click the green pencil icon beside the *API User Key*
.. In the *Set Custom User Key* modal dialog, enter:
+
[subs="attributes+"]
----
test
----
.. Select *Set Custom Key*.

[type=verification]
****
Use an HTTP client to invoke the route, for example:

----

> curl -d '{"name":"John"}' -H "Content-Type: application/json" -X POST "{route}/greeting?user_key=test"

Reviewer 3: `{route}` renders as {route} in walkthrough, and isn't replaced by actual route

----

Did the message "Hello from John" appear in your Slack channel?
****

[type=verificationFail]
Verify that you followed each step in the procedure above.  If you are still having issues, contact your administrator.

> === Invoking the API

Reviewer 3: This section seems to duplicate the previous one:

![Screenshot from 2019-08-22 15-22-57](https://user-images.githubusercontent.com/12085451/63522725-d36f6980-c4f0-11e9-814e-3d4f0645c3d8.png)
